eip_operator/src/controller/pod.rs

```diff
@@ -33,60 +33,45 @@ impl k8s_controller::Context for Context {
-
-        let node = node_api.get(node_name).await?;
-
-        let provider_id = node.provider_id().ok_or(Error::MissingProviderId)?;
-        let instance_id = provider_id
-            .rsplit_once('/')
-            .ok_or(Error::MalformedProviderId)?
-            .1;
-
-        let eni_id = match get_eni_id_from_annotation(pod) {
-            Some(eni_id) => eni_id,
-            None => {
-                let instance_description =
-                    crate::aws::describe_instance(&self.ec2_client, instance_id).await?;
-
-                crate::aws::get_eni_from_private_ip(&instance_description, pod_ip)
-                    .ok_or(Error::NoInterfaceWithThatIp)?
-            }
-        };
-
-        let all_eips = eip_api.list(&ListParams::default()).await?.items;
-        let eip = all_eips
-            .into_iter()
-            .find(|eip| eip.matches_pod(name))
-            .ok_or_else(|| Error::NoEipResourceWithThatPodName(name.to_owned()))?;
-        let eip_name = eip.name().ok_or(Error::MissingEipName)?;
-        let allocation_id = eip.allocation_id().ok_or(Error::MissingAllocationId)?;
-        let eip_description = crate::aws::describe_address(&self.ec2_client, allocation_id)
+        let eips = eip_api
+            .list(&ListParams::default())
             .await?
-            .addresses
-            .ok_or(Error::MissingAddresses)?
-            .swap_remove(0);
-        let public_ip = eip_description.public_ip.ok_or(Error::MissingPublicIp)?;
-        if eip_description.network_interface_id != Some(eni_id.to_owned())
-            || eip_description.private_ip_address != Some(pod_ip.to_owned())
-        {
-            crate::aws::associate_eip(&self.ec2_client, allocation_id, &eni_id, pod_ip).await?;
+            .items;
+        if let Some(eip) = eips.iter().find(|eip| eip.associated_with_pod(pod))  {
+            if ! eip.matches_pod_name(&pod.name_unchecked()) {
+                info!("Pod {} is attatched to EIP {}, but should not be, asking eip to reconcile, and rescheduling reconciliation", pod.name_unchecked(), eip.name_unchecked());
+                crate::eip::trigger_reconciliation(&eip_api, &eip.name_unchecked()).await?;
+                return Ok(Some(Action::requeue(Duration::from_secs(2))));
+            } else {
+                info!("Pod {} is already attatched to EIP {}, setting up DNS target annotation", pod.name_unchecked(), eip.name_unchecked());
+                let allocation_id = eip.allocation_id().ok_or(Error::MissingAllocationId)?;
+                let eip_description = crate::aws::describe_address(&self.ec2_client, allocation_id)
+                    .await?
+                    .addresses
+                    .ok_or(Error::MissingAddresses)?
+                    .swap_remove(0);
+                let public_ip = eip_description.public_ip.ok_or(Error::MissingPublicIp)?;
+                add_dns_target_annotation(&pod_api, name, &public_ip, allocation_id).await?;
```

Comment:
@alex-hunt-materialize 
There is a previously unhandled case here where `add_dns_target_annotation` only occurs when the EIP is added to a pod, if the eip no longer matches a the pod it is associated with we did not remove this annotation. My inclination is to let this one slip in this PR. I don't think it's likely we encounter this since we're not modifying EIPs this way and likely don't need to support pods much longer.